### PR TITLE
Allow u/npc_spawn_monster to pick randomly from a group for count > 1

### LINKED ISSUE #70812: Make u_spawn_monster works properly with mongroups
### Is your feature request related to a problem? Please describe.

U_spawn_monster still unable to spawn multiple monsters out of group, always spawning a single one, including using `"u_spawn_monster": ""`, that should pick random critter around and spawn it's copy

### Solution you would like.

Ability to pick do you want to spawn 50 copies of a single monster, or 50 different monsters 

### Describe alternatives you have considered.

_No response_

### Additional context

I tried multiple approaches, none of them was successful
image

### PR DESCRIPTION
**Pull request by Procyonae**
#### Summary
None

#### Purpose of change

Fixes #70812

#### Describe the solution

Adds a new field "single_type" that defaults to false but can be defined as true to replicate the old behaviour, I couldn't find any existing EoCs that want appear to want this to occur though (have a couple of ideas how it could be used tho)
Simplifies existing EoCs that were using the workaround of multiple 1 count EoCs
Updates documentation

#### Testing

Tested using the below EoCs with expected results
```
  {
    "type": "effect_on_condition",
    "id": "hallucination_swarm_testing",
    "effect": [
      { "u_spawn_monster": "", "hallucination_count": 20, "target_range": 50, "lifespan": [ "2 hours", "9 hours" ] }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "hallucination_swarm_testing_2",
    "effect": [
      { "u_spawn_monster": "", "hallucination_count": 20, "target_range": 50, "lifespan": [ "2 hours", "9 hours" ], "single_target": true }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "hallucination_swarm_testing_3",
    "effect": [
      { "u_spawn_monster": "GROUP_PET_DOGS", "real_count": 20, "group": true }
    ]
  },
  {
    "type": "effect_on_condition",
    "id": "hallucination_swarm_testing_4",
    "effect": [
      { "u_spawn_monster": "GROUP_PET_DOGS", "real_count": 20, "group": true, "single_target": true }
    ]
  }
```
If anyone can think of anything else I should test comment below

### CODE REVIEW

File: src/npctalk.cpp [hunk 1]
```diff
@@ -5585,29 +5586,63 @@ void talk_effect_fun_t::set_spawn_monster( const JsonObject &jo, std::string_vie
...
+                    bool valid_target = get_player_character().attitude_to( critter ) == Creature::Attitude::HOSTILE;
+                    return not_self && in_range && valid_target;
+                } );
+                if( copy == nullptr ) {
+                    run_eoc_vector( false_eocs, d );
+                    return;
+                }
+                target_monster = *copy->as_monster();
+            } else {
+                // Grab all hostile creatures to create hallucinations or copies of
```
Comment (GuardianDll) [inline src/npctalk.cpp:5622]: `u_spawn_monster` is not strictly about hallucinations

Comment (Procyonae) [reply src/npctalk.cpp:5622]: Ik, that part is only if using "u/"npc_spawn_monster": "" and specifies hallucinations or copies not just hallucinations

Comment (Procyonae) [inline src/npctalk.cpp:5622]:
```suggestion
                // Find all hostile creatures in range to be used to create hallucinations or copies of
```

File: src/npctalk.cpp [hunk 2]
```diff
@@ -5585,29 +5586,63 @@ void talk_effect_fun_t::set_spawn_monster( const JsonObject &jo, std::string_vie
...
-                bool not_self = get_player_character().pos() != critter.pos();
-                bool in_range = std::round( rl_dist_exact( get_player_character().pos(), critter.pos() ) ) <= target_range;
-                bool valid_target = get_player_character().attitude_to( critter ) == Creature::Attitude::HOSTILE;
-                return not_self && in_range && valid_target;
-            } );
-            if( copy == nullptr ) {
-                run_eoc_vector( false_eocs, d );
-                return;
+            if( single_target ) {
+                // Grab a random nearby hostile creature to create a hallucination or copy of
```
Comment (Procyonae) [inline src/npctalk.cpp:5609]:
```suggestion
                // Find a hostile creature in range to be used to create a hallucination or a copy of
```

File: data/json/effects_on_condition/npc_eocs/hallucination_eocs.json [hunk 3]
```diff
@@ -175,26 +175,31 @@
     "type": "effect_on_condition",
     "id": "creature_hallucinations",
     "//TODO": "ideally checks is the tile visible, and then spawns hallu where player can't see it",
-    "//": "if monsters are nearby, 3/4 chance to get copy of random monster nearby and run again (with 1/2 prob) and 1/4 chance to spawns one of YOUR_FEARS. If not, spawns one of YOUR_FEARS",
-    "effect": [
+    "//": "If monsters are nearby, 3/4 chance to spawn some hallucinations of random monsters nearby and run again half the time, 1/4 chance to spawn some of YOUR_FEARS. If not, spawns some of YOUR_FEARS",
+    "condition": { "and": [ { "math": [ "rand(3)", "==", "3" ] }, { "not": { "math": [ "u_monsters_nearby()", ">", "1" ] } } ] },
```
Comment (GuardianDll) [inline data/json/effects_on_condition/npc_eocs/hallucination_eocs.json:179]:
```suggestion
    "condition": { "and": [ { "math": [ "rand(3)", "==", "3" ] }, { "math": [ "u_monsters_nearby()", "==", "0" ] } ] },
```
I don't know why i made it in this way before

**Review by GuardianDll — COMMENTED** [review]
Looks good for me
Didn't know `default` can be used with math

File: data/json/effects_on_condition/npc_eocs/hallucination_eocs.json (hunk 3 above)
Comment (Procyonae) [reply data/json/effects_on_condition/npc_eocs/hallucination_eocs.json]: I figured maybe the greater than 1 was if it returns the player or something odd, I'll switch the effect and false_effect around to make more sense

Idk if the default there actually does anything bc if wave_spawn_number() fails to return anything I'd of thought there would be an error but I didn't look up the implementation

File: data/json/effects_on_condition/npc_eocs/hallucination_eocs.json [hunk 4]
```diff
@@ -209,19 +214,11 @@
...
-            "target_range": 50,
-            "max_radius": 50,
-            "lifespan": [ "2 hours", "9 hours" ]
-          },
-          { "math": [ "u_i", "++" ] },
-          { "run_eocs": "hallucination_swarm" }
-        ],
-        "else": { "math": [ "u_i", "=", "0" ] }
+        "u_spawn_monster": "",
+        "hallucination_count": [ 50, 100 ],
```
Comment (Procyonae) [inline data/json/effects_on_condition/npc_eocs/hallucination_eocs.json:218]:
```suggestion
        "hallucination_count": [ 50, 100 ],
```
This does result in a higher likelihood of 75+ hallucinations than the way it was working before if that's not desired

Comment (GuardianDll) [reply data/json/effects_on_condition/npc_eocs/hallucination_eocs.json:218]: No, it's fine

**Pull request merged**